src/Nethermind/Nethermind.Init/Steps/Migrations/TotalDifficultyFixMigration.cs

```diff
@@ -0,0 +1,89 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Nethermind.Api;
+using Nethermind.Blockchain;
+using Nethermind.Blockchain.Synchronization;
+using Nethermind.Core;
+using Nethermind.Int256;
+using Nethermind.Logging;
+using Nethermind.State.Repositories;
+
+namespace Nethermind.Init.Steps.Migrations;
+
+public class TotalDifficultyFixMigration : IDatabaseMigration
+{
+    private readonly IApiWithNetwork _api;
+    private readonly ILogger _logger;
+    private readonly ISyncConfig _syncConfig;
+
+    private Task? _fixTask;
+    private CancellationTokenSource? _cancellationTokenSource;
+
+    public TotalDifficultyFixMigration(IApiWithNetwork api, ISyncConfig syncConfig)
+    {
+        _api = api;
+        _logger = _api.LogManager.GetClassLogger();
+        _syncConfig = syncConfig;
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        _cancellationTokenSource?.Cancel();
+        await (_fixTask ?? Task.CompletedTask);
+    }
+
+    public void Run()
+    {
+        if (_syncConfig.FixTotalDifficulty)
+        {
+            _cancellationTokenSource = new CancellationTokenSource();
+            CancellationToken token = _cancellationTokenSource.Token;
+
+            _fixTask = Task.Run(() => RunMigration(_syncConfig.FixTotalDifficultyStartingBlock, _syncConfig.FixTotalDifficultyLastBlock, token), token)
+                .ContinueWith(x =>
+                    {
+                        if (x.IsFaulted && _logger.IsError)
+                        {
+                            _logger.Error($"Failed to finish TotalDifficultyFixMigration {x.Exception!.Message}");
+                        }
+                    });
+        }
+    }
+
+    private void RunMigration(long startingBlock, long? lastBlock, CancellationToken cancellationToken)
+    {
+        IChainLevelInfoRepository chainLevelInfoRepository = _api.ChainLevelInfoRepository ?? throw new ArgumentNullException(nameof(_api.ChainLevelInfoRepository));
+        IBlockTree blockTree = _api.BlockTree ?? throw new ArgumentNullException(nameof(_api.BlockTree));
```

Comment:
If migration is only using BlockTree and ChainLevelInfoHistory, it could accept these interfaces in constructor and perform argument check earlier.